sandbox/reflection/struct.h

```diff
@@ -0,0 +1,192 @@
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(F&& f, bricks::reflection::Index<bricks::reflection::FieldValue, idx>) {            \
+    f(name);                                                                                                  \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(                                                                                    \
+      F&& f, bricks::reflection::Index<bricks::reflection::FieldNameAndImmutableValueReference, idx>) const { \
+    f(#name, name);                                                                                           \
+  }                                                                                                           \
+  template <class F>                                                                                          \
+  void CURRENT_REFLECTION(                                                                                    \
+      F&& f, bricks::reflection::Index<bricks::reflection::FieldNameAndMutableValueReference, idx>) {         \
+    f(#name, name);                                                                                           \
+  }
+
+#define CURRENT_STRUCT_IS_VALID(s) \
+  bricks::reflection::CurrentStructFieldsConsistency<s, bricks::reflection::FieldCounter<s>::value - 1>::Check()
+
+namespace bricks {
+namespace reflection {
+
+template <typename T>
+struct FieldCounter {
+  static_assert(std::is_base_of<CurrentBaseType, T>::value,
+                "Template argument must be derived from `CurrentBaseType`.");
+  enum {
+    value = (sizeof(typename T::template CURRENT_REFLECTION_HELPER<T>::field_count_struct) / sizeof(long long))
+  };
+};
+
+template <int...>
+struct indexes {};
+template <int X, int... XS>
+struct indexes_generator : indexes_generator<X - 1, X - 1, XS...> {};
+template <int... XS>
+struct indexes_generator<0, XS...> {
+  typedef indexes<XS...> type;
+};
+
+template <int N>
+using gen_indexes = typename indexes_generator<N>::type;
+
+template <typename T, typename INDEX_TYPE>
+struct EnumFields {
+  static_assert(std::is_base_of<CurrentBaseType, T>::value,
+                "Template argument must be derived from `CurrentBaseType`.");
+  typedef gen_indexes<FieldCounter<T>::value> NUM_INDEXES;
+
+  template <typename F>
+  void operator()(F&& f) {
+    static NUM_INDEXES all_indexes;
+    Impl(std::move(f), all_indexes);
+  }
+
+ private:
+  template <typename F, int N, int... NS>
+  void Impl(F&& f, indexes<N, NS...>) {
+    static indexes<NS...> remaining_indexes;
+    T::CURRENT_REFLECTION(std::forward<F>(f), Index<INDEX_TYPE, N>());
+    Impl(std::move(f), remaining_indexes);
```

Comment:
Actually, I take it back -- the move semantics is wrong here, which also means the forward semantics is wrong too. What do you pass in here, a lambda or a real object? If only real objects, perhaps just `const F& f` is safer.